File: js/listaencadeada/listaenc.js
const prompt = require('prompt-sync')();

let lista = [];
let palavraLida = "";

let primeiro = null;

const criaNo = (p) => { 
  let obj = { palavra: p, prox: null };
  return(obj)
}

function lerPalavra() {
  const name = prompt('Digite uma palavra: ');
  return name;
}

function exibeLista() {
  let p = primeiro;
  while (p !== null) {
    console.log(lista[p].palavra);
    p = lista[p].prox;
  }
}

function inserePalavra(palavra) {
  let no = criaNo(palavra);  
  lista.push(no);

  let inserido = lista.length - 1;
  let atual, anterior;

  if (primeiro === null) {
    primeiro = inserido;
    atual = primeiro;    
  } else {
    atual = primeiro;
    while (atual < inserido) {
      if (lista[inserido].palavra < lista[atual].palavra) {
        if (atual === primeiro) {
          lista[inserido].prox = primeiro;
          primeiro = inserido;
        } else {
          lista[inserido].prox = atual;
          lista[anterior].prox = inserido;
        }
        break; 
      } else {
        if (lista[atual].prox === null) {
          lista[atual].prox = inserido;
          break; 
        } else {
          anterior = atual;
          atual = lista[atual].prox;
        }
      }     
    }    
  }   
}

while (true) {
  palavraLida = lerPalavra(); 

  if (palavraLida === "FIM") {
    break;
  }

  inserePalavra(palavraLida);

  console.log("LISTA: ", lista);
  exibeLista();
}

File: ruby/listaencadeada/listaenc.rb
def escolha_usuario
  puts "Digite uma palavra ou FIM"
  prompt = gets.chomp
  puts "\n"
  prompt
end

def node(p)
  {
    palavra: p,
    proximo: nil
  }
end

def exibe_lista(primeiro)
  atual = primeiro
  lista = []
  while atual
    lista << atual[:palavra]
    atual = atual[:proximo]
  end
  puts 'Lista Ordenada'
  puts lista.join("\n")
  puts "\n"
end

primeiro = nil

until (resposta = escolha_usuario) == 'FIM' do
  inserido = node(resposta)
  if primeiro.nil?
    primeiro = inserido
  elsif inserido[:palavra] < primeiro[:palavra]
    inserido[:proximo] = primeiro
    primeiro = inserido
  else
    if primeiro[:proximo].nil?
      primeiro[:proximo] = inserido
    else
      corrente = primeiro[:proximo]
      anterior = primeiro
      until inserido[:palavra] < corrente[:palavra] do
        if corrente[:proximo].nil?
          anterior = nil
          break
        end
        anterior = corrente
        corrente = corrente[:proximo]
      end
      anterior.nil? ? corrente[:proximo] = inserido : anterior[:proximo] = inserido
      inserido[:proximo] = corrente unless anterior.nil?
    end

  end
  exibe_lista(primeiro)
  puts "\n"
end
